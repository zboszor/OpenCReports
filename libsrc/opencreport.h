/*
 * OpenCReports main header
 * Copyright (C) 2019 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */
#ifndef _OPENCREPORT_H_
#define _OPENCREPORT_H_

#include <stdbool.h>
#include <stdint.h>
#include <time.h>
#include <mpfr.h>

/* Main report structure type */
struct opencreport;
typedef struct opencreport opencreport;

struct ocrpt_expr;
typedef struct ocrpt_expr ocrpt_expr;

enum ocrpt_result_type {
	/*
	 * Error in parsing or evaluation
	 */
	OCRPT_RESULT_ERROR,

	/*
	 * Constants
	 */
	OCRPT_RESULT_STRING,
	OCRPT_RESULT_NUMBER,
	OCRPT_RESULT_DATETIME
};

struct ocrpt_string {
	char *str;
	size_t allocated_len;
	size_t len;
};
typedef struct ocrpt_string ocrpt_string;

struct ocrpt_result {
	enum ocrpt_result_type type;
	/* Original lexer token or (computed) string value for expression */
	ocrpt_string *string;
	/* Converted numeric constant or computed numeric value for expression */
	mpfr_t number;
	bool number_initialized;
	bool string_owned;
	bool isnull;
	/* Datetime value */
	struct tm datetime;
};
typedef struct ocrpt_result ocrpt_result;

struct ocrpt_query;
typedef struct ocrpt_query ocrpt_query;

struct ocrpt_query_result {
	const char *name;
	bool name_allocated;
	ocrpt_result result;
};
typedef struct ocrpt_query_result ocrpt_query_result;

struct ocrpt_input {
	void (*describe)(ocrpt_query *, ocrpt_query_result **, int32_t *);
	void (*rewind)(ocrpt_query *);
	bool (*next)(ocrpt_query *);
	bool (*populate_result)(ocrpt_query *);
	bool (*isdone)(ocrpt_query *);
	void (*free)(ocrpt_query *);
};

typedef struct ocrpt_input ocrpt_input;

struct ocrpt_datasource;
typedef struct ocrpt_datasource ocrpt_datasource;

struct ocrpt_paper {
	const char *name;
	double width;
	double height;
};
typedef struct ocrpt_paper ocrpt_paper;

#if 0
struct report_line;
typedef struct report_line report_line;

struct ocrpt_report_part;
typedef struct ocrpt_report_part ocrpt_report_part;
#endif

#define OCRPT_MPFR_PRECISION_BITS	(256)

/*
 * Create a new empty OpenCReports structure
 */
opencreport *ocrpt_init(void);
/*
 * Free an OpenCReports structure
 */
void ocrpt_free(opencreport *o);
/*
 * Set MPFR numeric precision
 */
void ocrpt_set_numeric_precision_bits(opencreport *o, mpfr_prec_t prec);
/*
 * Set MPFR rounding mode
 */
void ocrpt_set_rounding_mode(opencreport *o, mpfr_rnd_t rndmode);

/********************************
 * Expression related functions *
 ********************************/

/*
 * Create an expression parse tree from an expression string
 */
ocrpt_expr *ocrpt_expr_parse(opencreport *o, const char *str, char **err);
/*
 * Optimize expression after parsing
 */
void ocrpt_expr_optimize(opencreport *o, ocrpt_expr *e);
/*
 * Resolve variable references in the expression
 */
void ocrpt_expr_resolve(opencreport *o, ocrpt_expr *e);
/*
 * Evaluate the expression, i.e. compute its ocrpt_result
 * It must be called after ocrpt_navigate_next(), see below.
 */
ocrpt_result *ocrpt_expr_eval(opencreport *o, ocrpt_expr *e);
/*
 * Print an expression on stdout. Good for unit testing.
 */
void ocrpt_expr_print(opencreport *o, ocrpt_expr *e);
/*
 * Print the result data. Good for unit testing.
 */
void ocrpt_expr_result_print(ocrpt_result *r);
/*
 * Count the number of expression nodes
 */
int ocrpt_expr_nodes(ocrpt_expr *e);
/*
 * Free an expression parse tree
 */
void ocrpt_free_expr(ocrpt_expr *e);

/****************************
 * Memory handling wrappers *
 ****************************/

typedef void *(*ocrpt_mem_malloc_t)(size_t);
typedef void *(*ocrpt_mem_realloc_t)(void *, size_t);
typedef void (*ocrpt_mem_free_t)(const void *);
typedef char *(*ocrpt_mem_strdup_t)(const char *);
typedef char *(*ocrpt_mem_strndup_t)(const char *, size_t);

extern ocrpt_mem_malloc_t ocrpt_mem_malloc0;
extern ocrpt_mem_realloc_t ocrpt_mem_realloc0;
extern ocrpt_mem_free_t ocrpt_mem_free0;
extern ocrpt_mem_strdup_t ocrpt_mem_strdup0;
extern ocrpt_mem_strndup_t ocrpt_mem_strndup0;

static inline void *ocrpt_mem_malloc(size_t sz) __attribute__((malloc,alloc_size(1)));
static inline void *ocrpt_mem_malloc(size_t sz) { return ocrpt_mem_malloc0(sz); }

static inline void *ocrpt_mem_realloc(void *ptr, size_t sz) __attribute__((alloc_size(2)));
static inline void *ocrpt_mem_realloc(void *ptr, size_t sz) { return ocrpt_mem_realloc0(ptr, sz); }

static inline void ocrpt_mem_free(const void *ptr) { ocrpt_mem_free0(ptr); }
static inline void ocrpt_strfree(const char *s) { ocrpt_mem_free0((const void *)s); }

static inline void *ocrpt_mem_strdup(const char *ptr) { return (ptr ? ocrpt_mem_strdup0(ptr) : NULL); }

static inline void *ocrpt_mem_strndup(const char *ptr, size_t sz) { return (ptr ? ocrpt_mem_strndup0(ptr, sz) : NULL); }

void ocrpt_mem_set_alloc_funcs(ocrpt_mem_malloc_t rmalloc,
								ocrpt_mem_realloc_t rrealloc,
								ocrpt_mem_free_t rfree,
								ocrpt_mem_strdup_t rstrdup,
								ocrpt_mem_strndup_t rstrndup);

/*
 * String related functions
 */
ocrpt_string *ocrpt_mem_string_new(const char *str, bool copy);
ocrpt_string *ocrpt_mem_string_new_with_len(const char *str, size_t len);
char *ocrpt_mem_string_free(ocrpt_string *string, bool free_str);
void ocrpt_mem_string_append_len(ocrpt_string *string, const char *str, const size_t len);
void ocrpt_mem_string_append(ocrpt_string *string, const char *str);
void ocrpt_mem_string_append_c(ocrpt_string *string, const char c);

/********************************
 * Paper size related functions *
 ********************************/

/*
 * System default paper name and size
 */
const ocrpt_paper *ocrpt_get_system_paper(void);
/*
 * Get paper size of the specified paper name
 */
const ocrpt_paper *ocrpt_get_paper_by_name(const char *paper);
/*
 * Set paper for the report using an ocrpt_paper structure
 * The contents of the structure is copied.
 */
void ocrpt_set_paper(opencreport *o, const ocrpt_paper *paper);
/*
 * Set paper for the report using a paper name
 * If the paper name is unknown, the system default paper is set.
 */
void ocrpt_set_paper_by_name(opencreport *o, const char *paper);
/*
 * Get the report's currently set paper
 */
const ocrpt_paper *ocrpt_get_paper(opencreport *o);
/*
 * Iterator for supported paper names and sizes
 * The returned paper structures are library-global
 * but the iterator is per report.
 */
const ocrpt_paper *ocrpt_paper_first(opencreport *o);
const ocrpt_paper *ocrpt_paper_next(opencreport *o);

/******************************************
 * Datasource and query related functions *
 ******************************************/

/*
 * Add a custom datasource not covered by
 * the currently implemented sources.
 */
ocrpt_datasource *ocrpt_add_datasource(opencreport *o, const char *source_name, const ocrpt_input *input);
/*
 * Find the datasource using its name
 */
ocrpt_datasource *ocrpt_find_datasource(opencreport *o, const char *source_name);
/*
 * Validate the datasource pointer against the report structure
 */
ocrpt_datasource *ocrpt_validate_datasource(opencreport *o, ocrpt_datasource *source);
/*
 * Add an array datasource
 *
 * Calling this is optional, as an array datasource called
 * "array" is automatically added to an opencreport structure.
 */
ocrpt_datasource *ocrpt_add_array_datasource(opencreport *o, const char *source_name);
/*
 * Add an array query using the datasource pointer
 */
ocrpt_query *ocrpt_add_array_query(opencreport *o, ocrpt_datasource *source,
									const char *name, void *array,
									int32_t rows, int32_t cols,
									const enum ocrpt_result_type *types);
/*
 * Add an array query using the datasource name
 */
ocrpt_query *ocrpt_add_array_query_as(opencreport *o, const char *source_name,
									const char *name, void *array,
									int32_t rows, int32_t cols,
									const enum ocrpt_result_type *types);
/*
 * Return the query result array and the number of columns in it
 *
 * It must be re-run for every new data source row since
 * the pointer is invalidated after ocrpt_navigate_next()
 */
ocrpt_query_result *ocrpt_query_get_result(ocrpt_query *q, int32_t *cols);
/*
 * Add follower query with a match function
 */
bool ocrpt_add_query_follower_n_to_1(opencreport *o,
									ocrpt_query *leader,
									ocrpt_query *follower,
									ocrpt_expr *match);
/*
 * Add follower query (runs side-by-side with "leader")
 */
bool ocrpt_add_query_follower(opencreport *o,
									ocrpt_query *leader,
									ocrpt_query *follower);
/*
 * Free a query and remove it from follower references
 */
void ocrpt_free_query(opencreport *o, ocrpt_query *q);
/*
 * Start query navigation from the beginning of the resultset
 */
void ocrpt_navigate_start(opencreport *o, ocrpt_query *q);
/*
 * Move to next row in the query resultset
 */
bool ocrpt_navigate_next(opencreport *o, ocrpt_query *q);

#if 0
/*
 * Create a report part
 */
ocrpt_report_part *ocrpt_create_report_part(void);
/*
 * Add report header to report part
 */
ocrpt_report_part_set_report_header(ocrpt_report_part *part, ...);
#endif
/*
 * Add XML report description from a file
 */
int ocrpt_add_report(opencreport *o, const char *filename);
/*
 * Add XML report description using a memory buffer (zero-terminated string)
 */
int ocrpt_add_report_from_buffer(opencreport *o, const char *buffer);
/*
 * Parse the reports added up to this point.
 * Allow no deviation from the OpenCReports DTD.
 */
int ocrpt_parse(opencreport *o);
/*
 * Parse the reports added up to this point.
 * Optionally allow deviation from the DTD.
 */
int ocrpt_parse2(opencreport *o, int allow_bad_xml);
/*
 * Execute the reports added up to this point.
 * Parse any unparsed report XMLs with non-strict checks.
 */
int ocrpt_execute(opencreport *o);

#endif /* _OPENCREPORT_H_ */
