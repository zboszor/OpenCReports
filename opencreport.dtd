<!ELEMENT OpenCReport (Datasources?,Queries?,(Part|Report)*)>
<!ELEMENT Datasources (Datasource+)>
<!ELEMENT Datasource EMPTY>
<!ATTLIST Datasource
	name CDATA #REQUIRED
	type CDATA #REQUIRED
	host CDATA #IMPLIED
	unix_socket CDATA #IMPLIED
	port CDATA #IMPLIED
	dbname CDATA #IMPLIED
	user CDATA #IMPLIED
	password CDATA #IMPLIED
	connstr CDATA #IMPLIED
	optionfile CDATA #IMPLIED
	group CDATA #IMPLIED >
<!ELEMENT Queries (Query)+>
<!ELEMENT Query (#PCDATA)>
<!ATTLIST Query
	name CDATA #REQUIRED
	datasource CDATA #REQUIRED
	value CDATA #IMPLIED
	rows CDATA #IMPLIED
	cols CDATA #IMPLIED
	coltypes CDATA #IMPLIED
	follower_for CDATA #IMPLIED
	follower_expr CDATA #IMPLIED >
<!ELEMENT Part (PageHeader|PageFooter|ReportHeader|ReportFooter|pr)*>
<!ATTLIST Part
	layout CDATA #IMPLIED
	pagesAcross CDATA #IMPLIED
	pages_across CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	font_size CDATA #IMPLIED
	orientation CDATA #IMPLIED
	topMargin CDATA #IMPLIED
	top_margin CDATA #IMPLIED
	leftMargin CDATA #IMPLIED
	left_margin CDATA #IMPLIED
	bottomMargin CDATA #IMPLIED
	bottom_margin CDATA #IMPLIED
	paperType CDATA #IMPLIED
	paper_type CDATA #IMPLIED
	iterations CDATA #IMPLIED
	suppressPageHeaderFirstPage CDATA #IMPLIED >
<!ELEMENT pr (pd)+>
<!ATTLIST pr
	layout CDATA #IMPLIED
	newpage CDATA #IMPLIED >
<!ELEMENT pd (Report*)>
<!ATTLIST pd
	width CDATA #IMPLIED
	height CDATA #IMPLIED
	border_width CDATA #IMPLIED
	border_color CDATA #IMPLIED >
<!ELEMENT Report ((Alternate|NoData)|(Breaks?,Variables?)|PageHeader|PageFooter|ReportHeader|ReportFooter|(Detail|Graph|Chart)*|xi:include)* >
<!ATTLIST Report
	pagesAcross CDATA #IMPLIED
	pages_across CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	font_size CDATA #IMPLIED
	orientation CDATA #IMPLIED
	topMargin CDATA #IMPLIED
	top_margin CDATA #IMPLIED
	leftMargin CDATA #IMPLIED
	left_margin CDATA #IMPLIED
	bottomMargin CDATA #IMPLIED
	bottom_margin CDATA #IMPLIED
	paperType CDATA #IMPLIED
	paper_type CDATA #IMPLIED
	iterations CDATA #IMPLIED
	suppressPageHeaderFirstPage CDATA #IMPLIED

	query CDATA #IMPLIED
	xmlns:xi CDATA #IMPLIED
	suppress CDATA #IMPLIED
	uniquerow CDATA #IMPLIED
	detail_columns CDATA #IMPLIED
	column_pad CDATA #IMPLIED >
<!ELEMENT xi:include EMPTY>
<!ATTLIST xi:include
	href CDATA #REQUIRED >
<!ELEMENT Alternate (NoData) >
<!ELEMENT NoData (Output) >
<!ELEMENT ReportHeader (Output)* >
<!ELEMENT PageHeader (Output)* >
<!ELEMENT ReportFooter (Output)* >
<!ELEMENT PageFooter (Output)* >
<!ELEMENT Output (Line | HorizontalLine | Image)* >
<!ELEMENT Line (field | literal)* >
<!ATTLIST Line
	fontSize CDATA #IMPLIED
	bold CDATA #IMPLIED
	suppress CDATA #IMPLIED
	color CDATA #IMPLIED
	bgcolor CDATA #IMPLIED >
<!ELEMENT HorizontalLine EMPTY>
<!ATTLIST HorizontalLine
	size CDATA #IMPLIED
	indent CDATA #IMPLIED
	length CDATA #IMPLIED
	fontSize CDATA #IMPLIED
	suppress CDATA #IMPLIED
	bgcolor CDATA #IMPLIED >
<!ELEMENT Image EMPTY >
<!ATTLIST Image
	value CDATA #REQUIRED
	type CDATA #REQUIRED
	width CDATA #REQUIRED
	height CDATA #REQUIRED >
<!ELEMENT field EMPTY>
<!ATTLIST field
	value CDATA #REQUIRED
	delayed CDATA #IMPLIED
	precalculate CDATA #IMPLIED
	width CDATA #IMPLIED
	align CDATA #IMPLIED
	format CDATA #IMPLIED
	bgcolor CDATA #IMPLIED
	color CDATA #IMPLIED
	link CDATA #IMPLIED
	maxlines CDATA #IMPLIED
	col CDATA #IMPLIED >
<!ELEMENT literal (#PCDATA) >
<!ATTLIST literal
	align CDATA #IMPLIED
	bgcolor CDATA #IMPLIED
	color CDATA #IMPLIED
	width CDATA #IMPLIED
	bold CDATA #IMPLIED
	italics CDATA #IMPLIED
	link CDATA #IMPLIED
	translate CDATA #IMPLIED
	col CDATA #IMPLIED >
<!ELEMENT Detail (FieldHeaders?, FieldDetails?) >
<!ELEMENT FieldHeaders (Output)* >
<!ELEMENT FieldDetails (Output)* >
<!ELEMENT Graph (Plot+) >
<!ATTLIST Graph
	name CDATA #IMPLIED
	type CDATA #REQUIRED
	subtype CDATA #REQUIRED
	width CDATA #IMPLIED
	height CDATA #IMPLIED
	bold_titles CDATA #IMPLIED
	title CDATA #IMPLIED
	legend_bg_color CDATA #IMPLIED
	legend_orientation CDATA #IMPLIED
	draw_x_line CDATA #IMPLIED
	draw_y_line CDATA #IMPLIED
	grid_color CDATA #IMPLIED
	x_axis_title CDATA #IMPLIED
	y_axis_title CDATA #IMPLIED
	y_axis_mod CDATA #IMPLIED
	y_axis_title_right CDATA #IMPLIED
	y_axis_decimals CDATA #IMPLIED
	y_axis_decimals_right CDATA #IMPLIED >
<!ELEMENT Plot EMPTY>
<!ATTLIST Plot
	axis CDATA #IMPLIED
	field CDATA #IMPLIED
	label CDATA #IMPLIED
	side CDATA #IMPLIED
	disabled CDATA #IMPLIED
	color CDATA #IMPLIED >
<!ELEMENT Breaks (Break)* >
<!ELEMENT Break (BreakHeader | BreakFooter | BreakFields)* >
<!ATTLIST Break
	name CDATA #REQUIRED
	headernewpage CDATA #IMPLIED
	suppressblank CDATA #IMPLIED
	newpage CDATA #IMPLIED >
<!ELEMENT BreakHeader (Output)* >
<!ELEMENT BreakFooter (Output)* >
<!ELEMENT BreakFields (BreakField)* >
<!ELEMENT BreakField EMPTY>
<!ATTLIST BreakField
	value CDATA #REQUIRED >
<!ELEMENT Variables (Variable)* >
<!ELEMENT Variable EMPTY >
<!ATTLIST Variable
	name CDATA #REQUIRED
	value CDATA #REQUIRED
	type (expression|count|sum|average|lowest|highest) "sum"
	resetonbreak CDATA #IMPLIED
	precalculate CDATA #IMPLIED
	delayed CDATA #IMPLIED >
