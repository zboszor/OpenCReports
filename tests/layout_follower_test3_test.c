/*
 * OpenCReports test
 * Copyright (C) 2019-2024 Zoltán Böszörményi <zboszor@gmail.com>
 * See COPYING.LGPLv3 in the toplevel directory.
 */

#include <stdio.h>
#include <stdlib.h>

#include <opencreport.h>
#include "test_common.h"

#define ROWS 5
#define COLS 5

const char *array[ROWS + 1][COLS] = {
	{ "first_name", "last_name", "color", "group", "breakfast" },
	{ "Bob", "Doan", "blue", "1", "Green Eggs And Spam I Am I Am" },
	{ "Eric", "Eburuschkin", "green", "1", "Green Eggs And Spam I Am I Am" },
	{ "Mike", "Roth", "yellow", "2", "Green Eggs And Spam I Am I Am" },
	{ "Bob", "Kratz", "pink", "2", "Green Eggs And Spam I Am I Am" },
	{ "Steve", "Tilden", "purple", "2", "Green Eggs And Spam I Am I Am" }
};

#define ROWS1 2
#define COLS1 1
const char *initials[ROWS1 + 1][COLS1] = {
	{ "initials" },
	{ "WRD" },
	{ "ERB" }
};

int main(int argc, char **argv) {
	opencreport *o = ocrpt_init();
	ocrpt_datasource *ds = ocrpt_datasource_add_array(o, "array");
	ocrpt_query *q = ocrpt_query_add_array(ds, "data", (const char **)array, ROWS, COLS, NULL, 0);
	ocrpt_query *q2 = ocrpt_query_add_array(ds, "more_data", (const char **)initials, ROWS1, COLS1, NULL, 0);

	ocrpt_query_add_follower(q, q2);

	if (!ocrpt_parse_xml(o, "follower.xml")) {
		printf("XML parse error\n");
		ocrpt_free(o);
		return 0;
	}

	ocrpt_set_output_format(o, argc >= 2 ? atoi(argv[1]) : OCRPT_OUTPUT_PDF);

	ocrpt_execute(o);

	ocrpt_spool(o);

	ocrpt_free(o);

	return 0;
}
