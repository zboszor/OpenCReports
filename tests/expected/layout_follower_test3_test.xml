<?xml version="1.0" encoding="utf-8"?>
<ocrpt>
  <part>
    <table>
      <tr>
        <td width="813.09">
          <report>
            <report_header>
              <output>
                <line>
                  <data width="401.00" font_point="11.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">Test Report (Normal) AND THE REPORT FORMAT IS: </data>
                  <data width="66.00" font_point="11.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">XML</data>
                  <data width="33.00" font_point="11.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">3</data>
                </line>
                <line>
                  <data width="813.09" font_point="11.00" font_face="Courier" bold="1" italics="0" align="left" bgcolor="#ffffff" color="#000000">Test Report (Bold)</data>
                </line>
                <line>
                  <data width="813.09" font_point="11.00" font_face="Courier" bold="0" italics="1" align="left" bgcolor="#ffffff" color="#000000">Test Report (Italics)</data>
                </line>
                <line>
                  <data width="813.09" font_point="11.00" font_face="Courier" bold="1" italics="1" align="left" bgcolor="#ffffff" color="#000000">Test Report (Bold and Italics)</data>
                </line>
                <line>
                  <data width="43.00" font_point="11.00" font_face="Courier" bold="1" italics="1" align="left" bgcolor="#ffffff" color="#000000">BOO: </data>
                  <data width="66.00" font_point="11.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">mike is a silly pants</data>
                </line>
              </output>
            </report_header>
            <field_headers>
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">First Name</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">Last Name</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">Color</data>
                </line>
              </output>
            </field_headers>
            <break_header name="break0">
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">1</data>
                </line>
              </output>
            </break_header>
            <field_details>
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Bob</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Doan</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">blue</data>
                  <data width="27.00" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">WRD</data>
                </line>
              </output>
            </field_details>
            <field_details>
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Eric</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Eburuschkin</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">green</data>
                  <data width="27.00" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">ERB</data>
                </line>
              </output>
            </field_details>
            <break_footer name="break0">
              <output>
                <line>
                  <data width="63.00" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">FOOTER</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">1</data>
                </line>
              </output>
            </break_footer>
            <break_header name="break0">
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">2</data>
                </line>
              </output>
            </break_header>
            <field_details>
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Mike</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Roth</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">yellow</data>
                  <data width="27.00" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">ERB</data>
                </line>
              </output>
            </field_details>
            <field_details>
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Bob</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Kratz</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">pink</data>
                  <data width="27.00" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">ERB</data>
                </line>
              </output>
            </field_details>
            <field_details>
              <output>
                <line>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Steve</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffff00" color="#000000">Tilden</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">purple</data>
                  <data width="27.00" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#e5e5e5" color="#000000">ERB</data>
                </line>
              </output>
            </field_details>
            <break_footer name="break0">
              <output>
                <line>
                  <data width="63.00" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">FOOTER</data>
                  <data width="108.01" font_point="9.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">2</data>
                </line>
              </output>
            </break_footer>
          </report>
        </td>
      </tr>
    </table>
    <part_page_footer>
      <output>
        <line>
          <data width="46.00" font_point="12.00" font_face="Courier" bold="0" italics="0" align="left" bgcolor="#ffffff" color="#000000">Page: </data>
          <data width="21.60" font_point="12.00" font_face="Courier" bold="0" italics="0" align="right" bgcolor="#ffffff" color="#000000">1</data>
        </line>
      </output>
    </part_page_footer>
  </part>
</ocrpt>
